File: Python/khoiluonghoachat.py
MOD = 1e9 + 7
def solve(n):
    dp = [[[0, 0] for _ in range(3)] for _ in range(n + 1)]
    dp[1][0][0] = dp[1][1][0] = 1
    for i in range(2, n + 1):
        dp[i][0][0] = sum(dp[i - 1][j][0] for j in range(3)) % MOD
        dp[i][0][1] = sum(dp[i - 1][j][k] for j in range(3) for k in range(2)) % MOD
        
        dp[i][1][0] = dp[i - 1][0][0]
        dp[i][1][1] = dp[i - 1][0][0] + dp[i - 1][0][1]
        if i > 2:
            dp[i][2][0] = dp[i - 1][1][0]
            dp[i][2][1] = dp[i - 1][1][0] + dp[i - 1][1][1]
    return sum(dp[n][j][k] for j in range(3) for k in range(2)) % MOD
print(solve(8))

File: Rust/src/khoiluonghoachat.rs
use std::io;

fn solve(s: &str) -> i32 {
    let mass: std::collections::HashMap<char, i32> = [('H', 1), ('C', 12), ('O', 16)].iter().cloned().collect();
    let mut stack: Vec<i32> = Vec::new();
    for char in s.chars() {
        if char.is_alphabetic() {
            stack.push(*mass.get(&char).unwrap());
        } else if char.is_numeric() {
            let temp = stack.pop().unwrap() * char.to_digit(10).unwrap() as i32;
            stack.push(temp);
        } else if char == '(' {
            stack.push(-1);
        } else if char == ')' {
            let mut temp: i32 = 0;
            while let Some(top) = stack.pop() {
                if top == -1 {
                    break;
                }
                temp += top;
            }
            stack.push(temp);
        }
    }

    stack.iter().sum()
}

fn main() {
    let mut input: String = String::new();
    io::stdin().read_line(&mut input).expect("Failed to read line");
    let t: i32 = input.trim().parse().expect("Invalid input");

    for _ in 0..t {
        let mut input: String = String::new();
        io::stdin().read_line(&mut input).expect("Failed to read line");
        println!("{}", solve(input.trim()));
    }
}